molecule/structure_checker2.h

```diff
@@ -0,0 +1,150 @@
+ * Copyright (C) from 2009 to Present EPAM Systems.
+ *
+ * This file is part of Indigo toolkit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
+
+#ifndef __structure_checker2__
+#define __structure_checker2__
+
+#include "base_c/defs.h"
+#include "base_cpp/exception.h"
+#include <string>
+#include <vector>
+#include <unordered_set>
+
+class IndigoObject;
+
+namespace indigo
+{
+
+    class BaseMolecule;
+
+    class DLLEXPORT StructureChecker2
+    {
+    public:
+        enum CheckTypeCode
+        {
+            CHECK_NONE = 0x00000000,         // Check none
+            CHECK_LOAD = 0x00000001,         // Check loading (correspondence some known format)
+            CHECK_VALENCE = 0x00000002,      // Check valence correctness
+            CHECK_RADICAL = 0x00000004,      // Check radicals existance
+            CHECK_PSEUDOATOM = 0x00000008,   // Check pseudoatoms existance
+            CHECK_STEREO = 0x00000010,       // Check strerochemistry description correctness
+            CHECK_QUERY = 0x00000020,        // Check query fetaures existance
+            CHECK_OVERLAP_ATOM = 0x00000040, // Check overlapping atoms existance
+            CHECK_OVERLAP_BOND = 0x00000080, // Check overlapping bonds existance
+            CHECK_RGROUP = 0x00000100,       // Check R-groups existance
+            CHECK_SGROUP = 0x00000200,       // Check S-groups existance
+            CHECK_TGROUP = 0x00000400,       // Check T-groups existance (SCSR features)
+            CHECK_CHIRALITY = 0x00000800,    // Check chirality feature correctness (including 3D source)
+            CHECK_CHIRAL_FLAG = 0x00001000,  // Check chiral flag existance (MOLFILE format)
+            CHECK_3D_COORD = 0x00002000,     // Check 3D coordinates existance
+            CHECK_CHARGE = 0x00004000,       // Check charged structure
+            CHECK_SALT = 0x00008000,         // Check possible salt structure
+            CHECK_AMBIGUOUS_H = 0x00010000,  // Check ambiguous H existance
+            CHECK_COORD = 0x00020000,        // Check coordinates existance
+            CHECK_V3000 = 0x00040000,        // Check v3000 format
+            CHECK_ALL = -1                   // Check all features (default)
+        };
+
+        static constexpr const char* checkTypeName[] = {"none",         "load",         "valence", "radical",    "pseudoatom", "stereo",    "query",
```

Comment:
Better use a map from the beginning.

Reply:
In addition, 
* if these literals are subject of interface, please expose them as named strings (or macros);
* if these literals are internal matter, please hide the initialization to cpp file.

Reply:
Mostly, these literals are here for self-documenting: they should be used in host language (J, Py etc.) and never in C/C++ code. C/C++ should use enum constants. If consts or macroses look clearer for these strings, OK, I'll rewrite it.

In the current code this is unclear thing: Python uses strings but which stings are available and how they map to the enum can be investigated by reverse engineering only.

Using of map is possible but redundant here (see **check_params_from_string(const char* params)**).

Reply:
Can Python/Java catch the strings from C interface file as well as from DLL interface? Probably, not. In this case, it's unnecessary to expose the strings, but enough to mention them in the documentation comment.

Reply:
> mention them in the documentation comment.

OK, will do.

Reply:
But this makes possible an inconsistency between doc comment and code if one of them is changed and another one forgotten to update.

Reply:
Yes, this is possible in general with any documentation. But what are alternatives?
We could provide a function that returns a list of recognizable strings of check types, but the caller anyway needs to know the semantics from some documentation. It can only use such a list for additional check before call.

Reply:
>But what are alternatives?
Please, look the alternative commited: public text constants and private string array. Good style and they can't be changed separately. C++ programmers typically shouldn't use the constants but they can.

Reply:
Just noticed that original comments contained spelling errors. They are copied as is to the new comments.
Could you please check and fix spelling once you are changing this interface files?
There are mistypes: existance, fetaures, strerochemistry.

Reply:
Fixed, pushed